### ISSUE TITLE
Craft 3.1 Migration Failure: m190120_000000_fix_supertablecontent_tables

### ISSUE THREAD

**Issue by roberttolton** [engram-design/SuperTable#225]
### Description

Got the following error when updating Super Table:
```
Exception: Unable to save a Super Table field’s settings without knowing its content table.

Migration: verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables

Output:

> update in {{%fields}} ... done (time: 0.002s)
Exception: Unable to save a Super Table field’s settings without knowing its content table. (/vagrant/vendor/verbb/super-table/src/services/SuperTableService.php:566)
#0 /vagrant/vendor/verbb/super-table/src/migrations/m190120_000000_fix_supertablecontent_tables.php(118): verbb\supertable\services\SuperTableService->saveSettings(Object(verbb\supertable\fields\SuperTableField))
#1 /vagrant/vendor/verbb/super-table/src/migrations/m190120_000000_fix_supertablecontent_tables.php(93): verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables->_updateMatrixOrSuperTableSettings(Object(verbb\supertable\fields\SuperTableField))
#2 /vagrant/vendor/craftcms/cms/src/db/Migration.php(56): verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables->safeUp()
#3 /vagrant/vendor/craftcms/cms/src/db/MigrationManager.php(243): craft\db\Migration->up(true)
#4 /vagrant/vendor/craftcms/cms/src/db/MigrationManager.php(163): craft\db\MigrationManager->migrateUp(Object(verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables))
#5 /vagrant/vendor/craftcms/cms/src/services/Updates.php(223): craft\db\MigrationManager->up()
#6 /vagrant/vendor/craftcms/cms/src/controllers/BaseUpdaterController.php(445): craft\services\Updates->runMigrations(Array)
#7 /vagrant/vendor/craftcms/cms/src/controllers/UpdaterController.php(208): craft\controllers\BaseUpdaterController->runMigrations(Array, 'restore-db')
#8 [internal function]: craft\controllers\UpdaterController->actionMigrate()
#9 /vagrant/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#10 /vagrant/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#11 /vagrant/vendor/craftcms/cms/src/web/Controller.php(109): yii\base\Controller->runAction('migrate', Array)
#12 /vagrant/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction('migrate', Array)
#13 /vagrant/vendor/craftcms/cms/src/web/Application.php(297): yii\base\Module->runAction('updater/migrate', Array)
#14 /vagrant/vendor/craftcms/cms/src/web/Application.php(683): craft\web\Application->runAction('updater/migrate')
#15 /vagrant/vendor/craftcms/cms/src/web/Application.php(243): craft\web\Application->_processUpdateLogic(Object(craft\web\Request))
#16 /vagrant/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#17 /vagrant/web/index.php(43): yii\base\Application->run()
#18 {main}
```

After navigating back to the updates page, was met with an un-finished updates message, proceeded through and confirmed the 'someone else might be updating' message, and got back to the CP successfully. No idea if the migrations successfully ran or if there's now hidden gremlins lurking.

### Steps to reproduce

1. Run Craft 3.1 update including all updates released as of 2018-01-21
2. Run Super Table 2.1.4.2 update

### Additional info

- Plugin version: Updating to 2.1.4.2
- Craft version: 3.1.2.2

**Comment by engram-design** [comment]
@roberttolton So just to confirm, the migration eventually went through and you can access the CP all okay?

**Comment by roberttolton** [comment]
@engram-design Only after going to the updates page again (not refreshing the error page), but the CP seems functional. It's perhaps also worth mentioning that while I have Super Table installed in this site, I don't actually have any fields using it.

**Comment by magnessjo** [comment]
@engram-design, @roberttolton,

  I can also confirm this issue occurring when updating to 2.1.4.2.  It stems from the migration.  

  I do have fields using super table. Currently, I don't see any issues with any of the fields.  I am going to deploy to staging to see what issues occur but I won't push to production until the issue is resolved to be safe.  

  I will add any relevant errors from my deployment to staging.  

Thanks!

**Comment by magnessjo** [comment]
@engram-design, @roberttolton,

  Strike that.  Some data from my fields was removed due to the migration.  I am going to refresh my database from production and see if I can reproduce the error to provide more information.  

**Update--** After restoring my local from production and rolling back to previously installed composer packages:

  I updated one plugin at a time.  I started with Redactor, SEOmatic, and then I did Craft (3.1.2.2).  After the Craft update, all of the fields within my super table fields are blank.  My fields are generally a Matrix Field type with multiple block types that contain super table fields (let me know if that is confusing).    

**Second Update** After upgrading to Craft (3.1.2.2), I upgraded to Super Table 2.1.0.  The migration error did not occur and all the data loss doesn't appear to have occurred.  Next I upgraded to Super Table 2.1.4.  The migration error did not occur but the data loss did.

**Comment by engram-design** [comment]
@magnessjo You shouldn't be updating Craft 3.1 and Super Table independently, as you'll definitely get data loss, even if you can't see it - if you have any Super Table + Matrix field combinations. This is because there was a bug in Craft 3.1.0 and Super Table 2.0.14. So updating to Craft 3.1.2.2 will cause you some issues with Super Table (anything below 2.1.4.x).

If possible, you'll want to roll back to Craft 3.0.x, and update both Craft and Super Table to their latest versions.

**Comment by magnessjo** [comment]
@engram-design,

  Thank you for the explanation.  I did rollback to my original versions, refreshed the db, and remove the storage directory.  

I was presented with a table does not exist error when I did the update from from Craftcms (3.0.31) to (3.1.2.2) and Super Table (2.0.14) to (2.1.4.3).  Please see the list error below.  

```
TypeError: Argument 1 passed to craft\db\Connection::tableExists() must be of the type string, null given,

migrations/m190120_000000_fix_supertablecontent_tables.php on line 36
```

**Comment by louderthan10** [comment]
Getting the exact same error moving from Craft 3.0.37 and SuperTable 2.0.14 (updating them at the same time). Also to note, my SuperTable fields are inside of a Neo field, if that makes any difference.

**Comment by engram-design** [comment]
This should be fixed in [2.1.5](https://github.com/verbb/super-table/releases/tag/2.1.5), sorry for the issue and let me know if there are any further issues.

If you run into any issues after migration, be sure to run the new `http://mysite.com/actions/super-table/plugin/check-content-tables` action endpoint to check Super Table content tables.

**Comment by magnessjo** [comment]
@engram-design,

  Thank you for the update.  I modified the versions for super table and craft to the latest and attempted to do a system update.  I am happy to report that the previous error does not show anymore, but I am getting another error.  I've included the message from the migration.  

Would you like me to open another issue or do you feel this is still related?  


```
Database Exception: SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'stc_7_documents' already exists
The SQL being executed was: RENAME TABLE stc_d81d2360-8d51-4824-adea-26662c317b78_documents TO stc_7_documents

Migration: verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables
```

**Comment by engram-design** [comment]
@magnessjo Good to hear, and this issue should be fixed in [2.1.5.3](https://github.com/verbb/super-table/releases/tag/2.1.5.3).

Its also not a massive issue - you can also just plain delete `stc_d81d2360-8d51-4824-adea-26662c317b78_documents` without any data loss.

**Comment by magnessjo** [comment]
@engram-design,

  I reverted and did the upgrade using 2.1.5.3 and everything worked fine.  All the data is showing.  

  Thank you for your time!

**Comment by louderthan10** [comment]
@engram-design Nice! Works for me too. 🙌

**Comment by louderthan10** [comment]
I'm getting this issue with 2.1.12 upgrading from Craft 2 to Craft 3:

### Steps to reproduce

With Craft 2 version of site loaded in database:
1. Get updated Craft 3 templates/files/env config in place on server
2. Run 'composer install'
3. Go to /admin to perform Craft upgrade

### Additional info

Craft Updater output:
```

Unknown Property: Getting unknown property: craft\fields\MissingField::contentTable

Migration: verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables

Output:

> Trying to create missing content table for #259 ...
> Field mismatch craft\fields\MissingField is not a Super Table field ...
> Trying to create missing content table for #372 ...
> Field mismatch craft\fields\MissingField is not a Super Table field ...
> Trying to create missing content table for #379 ...
> Field mismatch craft\fields\MissingField is not a Super Table field ...
> Trying to create missing content table for #281 ...
> Field mismatch craft\fields\MissingField is not a Super Table field ...
> Trying to create missing content table for #352 ...
> Field mismatch craft\fields\MissingField is not a Super Table field ...
Exception: Getting unknown property: craft\fields\MissingField::contentTable (/vendor/yiisoft/yii2/base/Component.php:154)
#0 /vendor/verbb/super-table/src/migrations/m190120_000000_fix_supertablecontent_tables.php(219): yii\base\Component->__get('contentTable')
#1 /vendor/craftcms/cms/src/db/Migration.php(56): verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables->safeUp()
#2 /vendor/craftcms/cms/src/db/MigrationManager.php(243): craft\db\Migration->up(true)
#3 /vendor/craftcms/cms/src/db/MigrationManager.php(163): craft\db\MigrationManager->migrateUp(Object(verbb\supertable\migrations\m190120_000000_fix_supertablecontent_tables))
#4 /vendor/craftcms/cms/src/services/Updates.php(223): craft\db\MigrationManager->up()
#5 /vendor/craftcms/cms/src/controllers/BaseUpdaterController.php(521): craft\services\Updates->runMigrations(Array)
#6 /vendor/craftcms/cms/src/controllers/UpdaterController.php(209): craft\controllers\BaseUpdaterController->runMigrations(Array, 'restore-db')
#7 [internal function]: craft\controllers\UpdaterController->actionMigrate()
#8 /vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#9 /vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#10 /vendor/craftcms/cms/src/web/Controller.php(109): yii\base\Controller->runAction('migrate', Array)
#11 /vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction('migrate', Array)
#12 /vendor/craftcms/cms/src/web/Application.php(297): yii\base\Module->runAction('updater/migrate', Array)
#13 /vendor/craftcms/cms/src/web/Application.php(683): craft\web\Application->runAction('updater/migrate')
#14 /vendor/craftcms/cms/src/web/Application.php(243): craft\web\Application->_processUpdateLogic(Object(craft\web\Request))
#15 /vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest(Object(craft\web\Request))
#16 /public/index.php(21): yii\base\Application->run()
#17 /vendor/laravel/valet/server.php(151): require('/Users/trav/Web...')
#18 {main}

```
- Plugin version: 1.0.6 -> 2.1.12 (the version in composer.json)
- Craft version: 2.7.5 -> 3.1.14

**Comment by engram-design** [comment]
If you refresh the page (the database update page), re-run the migration, that should get rid of the error. I've been trying (unsuccessfully) to figure out this issue, as I believe it only occurs once, when going from Craft 2 > Craft 3, and potentially something to do with how many things have changed.

I know that's a bit of a rough solution - but would you be able to try?

**Comment by louderthan10** [comment]
@engram-design Yeah! That seems to work. Biggest worry is that something was breaking in the process. I'll let you know if I notice any issues. Thanks!

**Comment by engram-design** [comment]
@louderthan10 From my testing there shouldn't be anything amiss (and if there is, potentially another issue). It must be something to do with being unable to register Super Table as a plugin (component) - but if its running this migration, then it obviously sees Super Table as a plugin - a a bit of a conundrum. I'll keep you posted if there are any more developments.
